subsys/net/lib/http/http_server.c

```diff
@@ -0,0 +1,1038 @@
+{
+	ctx_client->client_fd = new_socket;
+	ctx_client->offset = 0;
+	ctx_client->server_state = HTTP_PREFACE_STATE;
+	for (int i = 0; i < MAX_STREAMS; i++) {
+		ctx_client->streams[i].stream_state = HTTP_STREAM_IDLE;
+		ctx_client->streams[i].stream_id = 0;
+	}
+}
+
+struct http_stream_ctx *find_http_stream_context(struct http_client_ctx *ctx_client,
+						   uint32_t stream_id)
+{
+	for (int i = 0; i < MAX_STREAMS; i++) {
+		if (ctx_client->streams[i].stream_id == stream_id) {
+			return &ctx_client->streams[i];
+		}
+	}
+	return NULL;
+}
+
+struct http_stream_ctx *allocate_http_stream_context(struct http_client_ctx *ctx_client,
+						       uint32_t stream_id)
+{
+	for (int i = 0; i < MAX_STREAMS; i++) {
+		if (ctx_client->streams[i].stream_state == HTTP_STREAM_IDLE) {
+			ctx_client->streams[i].stream_id = stream_id;
+			ctx_client->streams[i].stream_state = HTTP_STREAM_OPEN;
+			return &ctx_client->streams[i];
+		}
+	}
+	return NULL;
+}
+
+int handle_http_request(struct http_server_ctx *ctx_server, struct http_client_ctx *ctx_client,
+			int client_index)
+{
+	int ret = -EINVAL;
+
+	do {
+		switch (ctx_client->server_state) {
+		case HTTP_PREFACE_STATE:
+			ret = handle_http_preface(ctx_client);
+			break;
+		case HTTP_REQUEST_STATE:
+			ret = handle_http1_request(ctx_server, ctx_client, client_index);
+			break;
+		case HTTP_FRAME_HEADER_STATE:
+			ret = handle_http_frame_header(ctx_server, ctx_client, client_index);
+			break;
+		case HTTP_FRAME_HEADERS_STATE:
+			ret = handle_http_frame_headers(ctx_client);
+			break;
+		case HTTP_FRAME_CONTINUATION_STATE:
+			ret = handle_http_frame_continuation(ctx_client);
+			break;
+		case HTTP_FRAME_SETTINGS_STATE:
+			ret = handle_http_frame_settings(ctx_client);
+			break;
+		case HTTP_FRAME_WINDOW_UPDATE_STATE:
+			ret = handle_http_frame_window_update(ctx_client);
+			break;
+		case HTTP_FRAME_RST_STREAM_STATE:
+			ret = handle_http_frame_rst_frame(ctx_server, ctx_client, client_index);
+		case HTTP_FRAME_GOAWAY_STATE:
+			ret = handle_http_frame_goaway(ctx_server, ctx_client, client_index);
+			break;
+		case HTTP_FRAME_PRIORITY_STATE:
+			ret = handle_http_frame_priority(ctx_client);
+			break;
+		case HTTP_DONE_STATE:
+			ret = handle_http_done(ctx_server, ctx_client, client_index);
+			break;
+		default:
+			LOG_ERR("Unknown state.");
+		}
+
+	} while (ret == 0 && ctx_client->offset > 0);
+
+	return ret;
+}
+
+int handle_http_frame_header(struct http_server_ctx *ctx_server,
+			      struct http_client_ctx *ctx_client, int client_index)
+{
+	LOG_DBG("HTTP_FRAME_HEADER");
+
+	int parse_result = parse_http_frame_header(ctx_client);
+
+	if (parse_result == 0) {
+		return -EAGAIN;
+	} else if (parse_result == -1) {
+		return -EINVAL;
+	}
+
+	ctx_client->bytes_parsed_in_current_frame = 0;
+
+	switch (ctx_client->current_frame.type) {
+	case HTTP_HEADERS_FRAME:
+		enter_http_frame_headers_state(ctx_server, ctx_client, client_index);
```

Comment:
We should take a uniform appraoch here, the current mixup of `return enter_http_xxx_state()`; and `enter_http_xxx_state(); break;` doesn't look well. I suggest to do any processing needed in current state **before** the switch instruction, and follow the `return enter_http_xxx_state()` pattern.